packages/wallet/src/wallet.ts

```diff
@@ -168,92 +172,151 @@ export class DemexWallet {
+    const signingData = await this.getSigningData(txRequest);
+    const { address, evmBech32Address, messages, signer, signingClient, signOpts, handler } = signingData;
+
+    await this.checkReloadAccountState({ address, evmBech32Address });
+
+    const accountState: AccountState | undefined = this.accountStates[address];
+
+    if (!accountState) throw Error(`account not found: ${address}`)
+
+    const timeoutHeight = await this.determineTimeoutHeight(signer);
+
+    const { sequence, accountNumber } = accountState;
+
+    const _signOpts: SignTxOpts = {
+      ...signOpts,
+      explicitSignerData: {
+        timeoutHeight,
+        accountNumber,
+        sequence,
+        ...signOpts?.explicitSignerData,
+      },
+    };
+
+    const [account] = await signer.getAccounts();
+    const signedTx = await this.getSignedTx(address, messages, signingClient, account, _signOpts);
+    this.updateAccountState(address, { sequence });
+
+    return {
+      ...signingData,
+      signerAddress: address,
+      signedTx,
+      signOpts: _signOpts,
+      handler: { ...handler, requestId: sequence.toString() },
+    }
+  }
+
+  async signAndBroadcast(
+    messages: EncodeObject[],
+    signOpts?: SignTxOpts,
+    broadcastOpts?: BroadcastTxOpts
+  ): Promise<BroadcastTxResult> {
+    const promise = new Promise<BroadcastTxResult>((resolve, reject) => {
+      this.txSignManager.enqueue({
         messages,
-        signedTx,
         broadcastOpts,
-        signOpts: _signOpts,
-        handler: { resolve, reject, requestId: sequence.toString() },
+        signOpts,
+        handler: { resolve, reject },
       });
+    });
+
+    return promise;
+  }
+
+  private async signAndBroadcastTx(txRequest: SignTxRequest) {
+    try {
+      const broadcastTx = await this.signAndConstructBroadcastTxRequest(txRequest);
+      this.txDispatchManager.enqueue(broadcastTx);
     } catch (error) {
-      reject(error);
+      txRequest.handler.reject(error);
     }
   }
 
   public async getSignedTx(
     signerAddress: string,
     messages: readonly EncodeObject[],
-    sequence: number,
-    opts: Omit<SignTxOpts, "sequence">,
+    signingClient: SigningStargateClient,
+    account: AccountData,
+    opts: SignTxOpts,
   ): Promise<Tx.TxRaw> {
-    const { memo = "", accountNumber, explicitSignerData, feeDenom } = opts;
-    const signingClient = await this.getSigningStargateClient();
-    const [account] = await this.signer.getAccounts();
+    const { explicitSignerData, feeDenom, feeGranter, fee } = opts;
+    const { sequence, accountNumber, memo = "" } = explicitSignerData ?? {};
 
     let signature: StdSignature | null = null;
     try {
       const chainId = await this.getChainId();
       await callIgnoreError(() => this.onRequestSign?.(messages));
       const signerData: SignerData = {
-        accountNumber: accountNumber ?? this.accountInfo!.accountNumber,
+        accountNumber: accountNumber ?? 0,
```

Comment:
should this ever be 0? for signing arbitrary message?

Reply:
we should default this to 0 in the deconstructor at L306

Reply:
> should this ever be 0? for signing arbitrary message?

is more of to comply to signOpts, cant really find a cleaner way to do it at the moment